jupiterone/resource_rule.go

```diff
@@ -4,342 +4,627 @@ import (
+	var plan, state *RuleModel
 
-	if v, ok := d.GetOk("polling_interval"); ok {
-		questionRuleInstance.PollingInterval = v.(string)
-	}
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
 
-	if v, ok := d.GetOk("outputs"); ok {
-		questionRuleInstance.Outputs = interfaceSliceToStringSlice(v.([]interface{}))
+	if resp.Diagnostics.HasError() {
+		return
 	}
 
-	if v, ok := d.GetOk("operations"); ok {
-		ops := make([]client.RuleOperation, 0)
-		err := json.Unmarshal([]byte(v.(string)), &ops)
-		if err != nil {
-			return nil, err
-		}
-		questionRuleInstance.Operations = ops
+	if !reflect.DeepEqual(plan, state) {
+		resp.Diagnostics.Append(resp.Plan.SetAttribute(ctx, path.Root("version"),
+			types.Int64Unknown())...)
 	}
+}
 
-	if v, ok := d.GetOk("question"); ok {
-		v := v.([]interface{})
-		if len(v) == 0 {
-			questionRuleInstance.Question = map[string]interface{}{}
-		} else {
-			// Question MaxItems is 1. Enforced at schema level.
-			questionRuleInstance.Question = v[0].(map[string]interface{})
-		}
-	}
+// Create implements resource.ResourceWithConfigure
+func (r *QuestionRuleResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var data *RuleModel
 
-	if v, ok := d.GetOk("question_id"); ok {
-		questionRuleInstance.QuestionId = v.(string)
-	}
+	// Read Terraform plan data into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
 
-	if v, ok := d.GetOk("question_name"); ok {
-		questionRuleInstance.QuestionName = v.(string)
+	if resp.Diagnostics.HasError() {
+		return
 	}
 
-	if v, ok := d.GetOk("templates"); ok {
-		questionRuleInstance.Templates = v.(map[string]interface{})
+	// TODO: in future versions, make operations more explicit in the schema
+	// TODO: This map can probably be replaces by existing structs as well
+	rule, err := data.BuildQuestionRuleInstance()
+	if err != nil {
+		resp.Diagnostics.AddError("failed to build rule from configuration", err.Error())
+		return
 	}
 
-	if v, ok := d.GetOk("tags"); ok {
-		questionRuleInstance.Tags = interfaceSliceToStringSlice(v.([]interface{}))
+	rule, err = r.client.CreateQuestionRuleInstance(rule)
+	if err != nil {
+		resp.Diagnostics.AddError("failed to create rule", err.Error())
+		return
 	}
 
-	if v, ok := d.GetOk("version"); ok {
-		questionRuleInstance.Version = v.(int)
+	data.Id = types.StringValue(rule.Id)
+	data.Version = types.Int64Value(int64(rule.Version))
+
+	// TODO: This should _probably_ be done whenever values are read from the server
+	// and stored back to state, but version 0.5.0 didn't do this and relied
+	// solely on the suppress diff func, so leave this out for test/state
+	// compatibility until more improvements worth upgrading the state for are
+	// implemented
+	/* data.Operations, err = processOperationsState(rule.Operations)
+	if err != nil {
+		resp.Diagnostics.AddError("failed to save operations state", err.Error())
+		return
+	} */
```

Comment:
👍 I'm guessing using `jsonencode` would be a big help to this?

Reply:
> 👍 I'm guessing using `jsonencode` would be a big help to this?

Yes, I opened #139 to better describe what I think should be a straightforward implementation once the framework is in use.